File: csharp/Solitaire/Klondike/BoardValidator.cs
using Silnith.Game.Deck;
using System;
using System.Collections.Generic;

namespace Silnith.Game.Klondike
{
    /// <summary>
    /// A validator for the entire board in a game of Klondike solitaire.
    /// </summary>
    /// <remarks>
    /// <para>
    /// This validates that the foundation is properly ordered, that runs
    /// in each column conform to the rules, and that the total number of
    /// cards matches the expected size of a deck.
    /// </para>
    /// </remarks>
    public class BoardValidator : IValidator<Board>
    {
        private readonly int columnCount;
        private readonly int deckCount;
        private readonly ColumnValidator columnValidator;
        private readonly FoundationValidator foundationValidator;

        /// <summary>
        /// Constructs a new board validator.
        /// </summary>
        /// <param name="columnCount">The number of columns on the board.</param>
        /// <param name="deckCount">The number of cards in a complete deck.</param>
        /// <param name="columnValidator">A column validator.</param>
        /// <param name="foundationValidator">A foundation validator.</param>
        public BoardValidator(int columnCount, int deckCount, ColumnValidator columnValidator, FoundationValidator foundationValidator)
        {
            this.columnValidator = columnValidator ?? throw new ArgumentNullException(nameof(columnValidator));
            this.foundationValidator = foundationValidator ?? throw new ArgumentNullException(nameof(foundationValidator));
            this.deckCount = deckCount;
            this.columnCount = columnCount;
        }

        /// <inheritdoc/>
        public void Validate(Board board)
        {
            foundationValidator.Validate(board.Foundation);
            if (board.Columns.Count != columnCount)
            {
                throw new ArgumentException("Board must have " + columnCount + " columns, instead has " + board.Columns.Count + ".");
            }
            foreach (Column column in board.Columns)
            {
                columnValidator.Validate(column);
            }
            if (board.StockPileIndex < 0)
            {
                throw new ArgumentException("Stock pile index must be non-negative.");
            }
            if (board.StockPileIndex > board.StockPile.Count)
            {
                throw new ArgumentException("Stock pile index must not be larger than the stock pile size.  Stock pile has "
                    + board.StockPile.Count + " cards, stock pile index is " + board.StockPileIndex + ".");
            }
            int numCards = board.StockPile.Count;
            foreach (IReadOnlyList<Card> cards in board.Foundation.Values)
            {
                numCards += cards.Count;
            }
            foreach (Column column in board.Columns)
            {
                numCards += column.GetCountOfFaceDownCards();
                numCards += column.GetCountOfFaceUpCards();
            }
            if (numCards != deckCount)
            {
                throw new ArgumentException("Must have " + deckCount + " cards on the board, instead has " + numCards + ".");
            }
        }
    }
}

File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/BoardValidator.java
package org.silnith.game.solitaire;

import java.util.List;
import java.util.Map;

import org.silnith.deck.Card;
import org.silnith.deck.Suit;
import org.silnith.game.Validator;


/**
 * A validator for the entire board in a game of Klondike solitaire.
 * 
 * <p>This validates that the foundation is properly ordered, that runs
 * in each column conform to the rules, and that the total number of
 * cards matches the expected size of a deck.</p>
 */
public class BoardValidator implements Validator<Board> {
    
    private final int numberOfColumns;
    
    private final int deckSize;
    
    private final ColumnValidator columnValidator;
    
    private final FoundationValidator foundationValidator;
    
    /**
     * Creates a new board validator.
     * 
     * @param numberOfColumns the number of columns on the board.  In a normal game, this is {@code 7}.
     * @param deckSize the size of the deck.  In a normal game, this is {@code 52}.
     * @param columnValidator a column validator
     * @param foundationValidator a foundation validator
     */
    public BoardValidator(final int numberOfColumns, final int deckSize, final ColumnValidator columnValidator,
            final FoundationValidator foundationValidator) {
        super();
        this.numberOfColumns = numberOfColumns;
        this.deckSize = deckSize;
        this.columnValidator = columnValidator;
        this.foundationValidator = foundationValidator;
    }
    
    @Override
    public void validate(final Board board) {
        if (board == null) {
            throw new IllegalArgumentException("Board cannot be null.");
        }
        final Map<Suit, List<Card>> foundation = board.getFoundation();
        foundationValidator.validate(foundation);
        final List<Column> columns = board.getColumns();
        if (columns == null) {
            throw new IllegalArgumentException("Board has no list of columns.");
        }
        if (columns.size() != numberOfColumns) {
            throw new IllegalArgumentException(
                    "Board must have " + numberOfColumns + " columns, instead has " + columns.size());
        }
        for (final Column column : columns) {
            columnValidator.validate(column);
        }
        final List<Card> stockPile = board.getStockPile();
        if (stockPile == null) {
            throw new IllegalArgumentException("Board stock pile cannot be null.");
        }
        final int stockPileIndex = board.getStockPileIndex();
        if (stockPileIndex < 0) {
            throw new IllegalArgumentException("Stock pile index must be non-negative, is " + stockPileIndex + ".");
        }
        if (stockPileIndex > stockPile.size()) {
            throw new IllegalArgumentException("Stock pile index must not be larger than the stock pile size.  Stock pile has "
                    + stockPile.size() + " cards, stock pile index is " + stockPileIndex + ".");
        }
        // count the cards
        int numCards = stockPile.size();
        for (final Suit suit : Suit.values()) {
            numCards += foundation.get(suit).size();
        }
        for (final Column column : columns) {
            numCards += column.getNumberOfFaceDownCards();
            numCards += column.getNumberOfFaceUpCards();
        }
        if (numCards != deckSize) {
            throw new IllegalArgumentException(
                    "Must have " + deckSize + " cards on the board, instead has " + numCards + ".");
        }
    }
    
}
